File: src/business/publish-from-spu/qualification-image-normalizer.ts
import { execFile } from "node:child_process";
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { promisify } from "node:util";
import { getPythonCommand } from "../../utils/platform.js";
import {
  resolveQualificationImageResize,
  verifyNormalizedQualificationImage
} from "./qualification-image-rules.js";

const execFileAsync = promisify(execFile);
const NORMALIZER_SCRIPT = path.join(
  process.cwd(),
  "src",
  "business",
  "publish-from-spu",
  "qualification-image-normalizer.py"
);

type ImageProbe = {
  ok: boolean;
  width: number;
  height: number;
  format?: string;
  error?: string;
};

export type PreparedQualificationImages = {
  files: string[];
  entries: Array<{
    sourceFile: string;
    outputFile: string;
    action: "reuse" | "resize";
    sourceDimensions: { width: number; height: number };
    outputDimensions: { width: number; height: number };
  }>;
};

function fileHash(file: string): string {
  return crypto.createHash("sha256").update(fs.readFileSync(file)).digest("hex");
}

function parsePayload(output: string, context: string): ImageProbe {
  const text = output.trim();
  const match = text.match(/(\{[\s\S]*\})\s*$/);
  if (!match) {
    throw new Error(`${context}: image processor returned no JSON. output=${text.slice(-500)}`);
  }
  const payload = JSON.parse(match[1]) as ImageProbe;
  if (!payload.ok) {
    throw new Error(`${context}: ${payload.error || "image processor failed"}`);
  }
  return payload;
}

async function runProcessor(args: string[], context: string): Promise<ImageProbe> {
  try {
    const { stdout, stderr } = await execFileAsync(
      getPythonCommand(),
      ["-X", "utf8", NORMALIZER_SCRIPT, ...args],
      {
        windowsHide: true,
        maxBuffer: 1024 * 1024 * 8,
        encoding: "utf8",
        env: { ...process.env, PYTHONIOENCODING: "utf-8" }
      }
    );
    return parsePayload(`${stdout}\n${stderr}`, context);
  } catch (error) {
    const stdout = String((error as { stdout?: string }).stdout || "");
    const stderr = String((error as { stderr?: string }).stderr || "");
    const details = `${stdout}\n${stderr}`.trim();
    if (details) {
      return parsePayload(details, context);
    }
    throw new Error(`${context}: ${error instanceof Error ? error.message : String(error)}`);
  }
}

async function probeImage(file: string): Promise<ImageProbe> {
  return runProcessor(["--input", file], `Qualification image dimension probe failed for ${path.basename(file)}`);
}

function outputPathFor(sourceFile: string, outputDir: string, hash: string): string {
  const extension = path.extname(sourceFile).toLowerCase() || ".png";
  const stem = path.basename(sourceFile, path.extname(sourceFile));
  return path.join(outputDir, `${stem}-${hash.slice(0, 8)}-max4900${extension}`);
}

export async function prepareQualificationImagesForUpload(options: {
  files: string[];
  outputDir: string;
}): Promise<PreparedQualificationImages> {
  const files: string[] = [];
  const entries: PreparedQualificationImages["entries"] = [];

  for (const sourceFile of options.files) {
    if (!fs.existsSync(sourceFile)) {
      throw new Error(`Qualification image source file was missing: ${sourceFile}`);
    }
    const sourceHash = fileHash(sourceFile);
    const sourceProbe = await probeImage(sourceFile);
    const decision = resolveQualificationImageResize({
      width: sourceProbe.width,
      height: sourceProbe.height
    });

    if (decision.action === "reuse") {
      files.push(sourceFile);
      entries.push({
        sourceFile,
        outputFile: sourceFile,
        action: "reuse",
        sourceDimensions: { width: sourceProbe.width, height: sourceProbe.height },
        outputDimensions: { width: sourceProbe.width, height: sourceProbe.height }
      });
      continue;
    }

    fs.mkdirSync(options.outputDir, { recursive: true });
    const outputFile = outputPathFor(sourceFile, options.outputDir, sourceHash);
    await runProcessor(
      [
        "--input",
        sourceFile,
        "--output",
        outputFile,
        "--width",
        String(decision.targetWidth),
        "--height",
        String(decision.targetHeight)
      ],
      `Oversized qualification image normalization failed for ${path.basename(sourceFile)} (${sourceProbe.width}x${sourceProbe.height})`
    );
    if (!fs.existsSync(outputFile)) {
      throw new Error(`Normalized qualification image output was missing: ${outputFile}`);
    }
    const outputProbe = await probeImage(outputFile);
    const verification = verifyNormalizedQualificationImage({
      width: outputProbe.width,
      height: outputProbe.height,
      targetWidth: decision.targetWidth,
      targetHeight: decision.targetHeight
    });
    if (!verification.passed) {
      throw new Error(`${verification.issue} source=${path.basename(sourceFile)}`);
    }
    if (fileHash(sourceFile) !== sourceHash) {
      throw new Error(`Qualification image source changed during normalization: ${sourceFile}`);
    }
    files.push(outputFile);
    entries.push({
      sourceFile,
      outputFile,
      action: "resize",
      sourceDimensions: { width: sourceProbe.width, height: sourceProbe.height },
      outputDimensions: { width: outputProbe.width, height: outputProbe.height }
    });
  }

  return { files, entries };
}

File: src/business/publish-from-spu/qualification-image-normalizer.py
import argparse
import json
import os
import sys

from PIL import Image, ImageOps


def probe(input_file: str) -> dict:
    with Image.open(input_file) as source:
        image = ImageOps.exif_transpose(source)
        return {
            "ok": True,
            "width": image.width,
            "height": image.height,
            "format": source.format or "",
        }


def resize(input_file: str, output_file: str, width: int, height: int) -> dict:
    os.makedirs(os.path.dirname(output_file), exist_ok=True)
    with Image.open(input_file) as source:
        image = ImageOps.exif_transpose(source)
        resized = image.resize((width, height), Image.Resampling.LANCZOS)
        extension = os.path.splitext(output_file)[1].lower()
        if extension in (".jpg", ".jpeg") and resized.mode not in ("RGB", "L"):
            resized = resized.convert("RGB")
        resized.save(output_file)
    return probe(output_file)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--input", required=True)
    parser.add_argument("--output")
    parser.add_argument("--width", type=int)
    parser.add_argument("--height", type=int)
    args = parser.parse_args()

    try:
        if args.output:
            if not args.width or not args.height or args.width <= 0 or args.height <= 0:
                raise ValueError("Resize mode requires positive --width and --height values.")
            payload = resize(args.input, args.output, args.width, args.height)
        else:
            payload = probe(args.input)
        print(json.dumps(payload, ensure_ascii=False))
        return 0
    except Exception as error:
        print(json.dumps({"ok": False, "error": str(error)}, ensure_ascii=False))
        return 1


if __name__ == "__main__":
    sys.exit(main())
